DM-28314: Initial JSON investigation

timj opened the pull request:
This implements the simplest possible JSON support.

```
json = ref.to_json()
ref = DatasetRef.from_json(json, universe=universe)
```

A universe is required although I imagine we could make it optional if we could have a support serialization package that could be queried for a default universe to use.

To implement these I've also created `to_simple` and `from_simple` pairs for `DatasetRef`, `DatasetType`, `DimensionGraph` and `DataCoordinate` that convert the object to a dict or list and then take that and convert it back to the object.

timj committed:
Add basic JSON support to DatasetType

timj committed:
Add JSON support for DatasetRef

timj commented:
@TallJimbo would you mind taking a quick look at this initial implementation before I continue on.

Thinking out loud about the client/server usage, it might be a good idea to add a `minimalist=True` argument to the `to_json` method to let it return something minimal for wire transport to a server. For example for a `DatasetRef` you in theory only need the `id` and the server can work out what you meant, and for `DatasetType` the name is sufficient. This would then need the `from_json()` methods to take an optional registry argument to let them do a registry lookup to create the full object in the server.  The server should return fully-specified results to the client otherwise that could result in additional registry lookups to the server to resolve everything later on.

So client Registry always uses `to_json(minimalist=True)`, server always used `from_json(registry=R)` and server always generates JSON with `to_json(minimalist=False)`. Or something like that.

TallJimbo reviewed (approved):
Looks great!  Sorry it took me so long to take a look.

I definitely like standardizing on `to_json` and `from_json`, and those delegating to `to_simple` and `from_simple` and the way you're handling the `DimensionUniverse` in those.

I'm trying to think of a better name to replace "simple" but haven't come up with one.  I'd have probably used "pod" for Plain Old Data, but as I think we've discussed I apparently think of that terminology as way more common that it actually is.

What do you think about a decorator or mixin to inject the `to_json` and `from_json` methods that delegate to `to_simple` and `from_simple`?  Seems like we should do something like that to make these mechanical, but that may not be compatible with universe handling and/or MyPy.

timj committed:
Support minimal JSON encoding for DatasetRef

This allows the object reconstruction to use registry to
construct the complete object. For a DatasetRef an id is
sufficient and for a DatasetType the name is sufficient.

timj committed:
Unify interface for to_simple/from_simple

so minimal is passed in even if not used and registry
is passed in to from_simple

timj committed:
Move to_json and from_json to separate module

This minimizes code duplication.

timj committed:
Add JSON serialization to Timespan

Currently uses the internal nanosecond form.

timj committed:
Add JSON serialization for DimensionElement

timj committed:
Add JSON support for DimensionRecord

TallJimbo commented on python/lsst/daf/butler/core/datasets/ref.py:
```diff
@@ -152,6 +159,108 @@ def __lt__(self, other: Any) -> bool:
+            # is the integer id so that can be used directly if it is
+            # resolved and if it is not a component DatasetRef
+            if not self.isComponent():
+                return self.id
```
Would it simplify the `from_simple` implementation much to return a dict with just the `id` key here?  If we someday add an `origin` term to the primary key, that'll be an easier transition.

And, if our primitives always use `Dict[str, Any]` as their "simple" representation (i.e. we go in a similar direction with `DatasetType`), we'll be able to more easily add a `Protocol` (thanks, Python 3.8) for the `to_simple`/`from_simple` interfaces and get a bit more type safety when we start using them.

timj replied:
We can always use a dict given that the origin will turn up at some point.

Not so sure about turning DatasetType into `{"name": nam}` but that could be done.

timj replied:
I've changed DatasetRef to always return a dict.

It's probably the case that switching to pydantic at some point would result in everything being a dict so maybe I should switch DatasetType over as well in the interests of consistency even though it adds 8 extra characters to all the serializations.

TallJimbo commented on python/lsst/daf/butler/core/datasets/ref.py:
```diff
@@ -152,6 +159,108 @@ def __lt__(self, other: Any) -> bool:
+        # Minimalist component will just specify component and id and
+        # require registry to reconstruct
+        if set(simple) == {"id", "component"}:
+            if registry is None:
```
It seems like it might be more efficient to swap these checks, and assume the input was produced with `minimal=True` if `registry is not None`.  I have no idea if that's the kind of "more efficient" that could matter down the road, but it does seem conceivable that we'll spend a lot of time in this code in the client/server butler.

timj replied:
I think the server is always going to pass in a registry parameter and the client is always going to call `from_simple` with a universe, but we can't enforce every client using the minimal parameter, so I always have to look at the content to see whether it is minimal or not.

TallJimbo commented on python/lsst/daf/butler/core/datasets/ref.py:
```diff
@@ -152,6 +159,108 @@ def __lt__(self, other: Any) -> bool:
+        if set(simple) == {"id", "component"}:
+            if registry is None:
+                raise ValueError("Registry is required to construct component DatasetRef from integer id")
+            parent_ref = registry.getDataset(simple["id"])
```
...but a bigger issue for performance is almost certainly the fact that this call isn't vectorized, and it'll be death-by-many-tiny-queries on the server whenever anyone sends a long list of minimal `DatasetRefs` over the wire.  We can of course work around that by not using the minimal form, but maybe this is a sign that we should be using `FakeDatasetRef` in more places and/or just reworking what a `DatasetRef` is supposed to be.  Not a problem for this ticket, I think, but potentially a big one for the client/server butler project (unless sending non-minimal DatasetRefs around ends up being totally fine).

Also note that if we started out with a `DatasetRef` such that `ref.dataId.hasRecords() is True`, that will not be the case after round-trip through minimal form (and changing that would be many-tiny-queries x20).  That's probably okay, it's something to look out for (and probably document).

timj replied:
I think we might want a specialist method in the server for processing the collection of JSON simple dataset refs. The server always has to send the full form to the client but the client can send the minimal form back to the server. We can tell up front in the server if we are processing a collection of them.

TallJimbo commented on python/lsst/daf/butler/core/datasets/type.py:
```diff
@@ -472,6 +474,94 @@ def _lookupNames(self) -> Tuple[LookupKey, ...]:
+            if registry is None:
+                raise ValueError(f"Unable to convert a DatasetType name '{simple}' to DatasetType"
+                                 " without a Registry")
+            datasetTypes = list(registry.queryDatasetTypes(simple))
```
I suspect this is going to be a fair bit slower than `Registry.getDatasetType`, and it should be impossible to get more than one result anyway; there's a unique constraint in the database, and the pattern-matching logic won't be in play unless the object passed in is a `re.Pattern` object, not a `str`.

Also, this minimal form should be safe from the many-tiny-queries problem that `DatasetRef` suffers from, at least for our current manager implementations, because those aggressively fetch all `DatasetTypes` from the database and stuff them in dict, so no DB calls should be needed here.  Hopefully that sort of aggressive fetching and/or caching scales up to the full science-user database...

timj replied:
I have no idea why I was using `query` and not `get` there...

TallJimbo commented on python/lsst/daf/butler/core/datasets/type.py:
```diff
@@ -472,6 +474,94 @@ def _lookupNames(self) -> Tuple[LookupKey, ...]:
+            Newly-constructed object.
+        """
+        if isinstance(simple, str):
+            if registry is None:
```
Same question as with `DatasetRef` as to whether checking for `registry is None` in the outer block would speed things up and/or simplify the code.

timj replied:
Is an "is None" test expensive? The instance test presumably costs more (as would a test for a dict with only "name" in it). The client is always going to call this with universe and the server is always going to call this with registry but I'm not sure we can rely on that to be true for all possible JSON serializations.

TallJimbo replied:
I would guess that `is None` boils down to a pointer comparison in C, so it's definitely going to be faster than `isinstance`, but I'd bet `isinstance(obj, str)` is similarly fast when the object actually is a `str` (I'd bet both cases are dominated by just per-statement interpreter overheads). It's the case where it's going to return `False` that I don't know about; it seems like it _should_ be possible for that to be quite fast, too, for `str` and other builtins, even if it might be much slower for user-defined classes.

TallJimbo commented on python/lsst/daf/butler/core/dimensions/_records.py:
```diff
@@ -175,6 +181,105 @@ def __reduce__(self) -> tuple:
+                if isinstance(v, lsst.sphgeom.Region):
+                    # Match YAML serialization
+                    mapping[k] = {"cls": f"lsst.sphgeom.{type(v).__name__}",
+                                  "encoded": v.encode().hex()}
```
I actually noticed the other day that the sphgeom class is already included in the encoded representation, and you can use `Region.decode` to handle a byte string produced by any derived class.  It's might be too late to be worth switching the YAML representation, but we could get away with just `bytes` for this one, if we wanted to.

timj replied:
I noticed too that the YAML form was storing the class and then not using it. It seems like I messed up somewhere as we iterated to the final form. On the other hand the `cls` here does make it obvious that it's a region although using a single key in a dict of `encoded_region` would also do that.

timj replied:
Changed it to just use the "encoded_region" key. It can't use bytes because json requires strings.

TallJimbo commented on python/lsst/daf/butler/core/json.py:
```diff
@@ -0,0 +1,91 @@
+    try:
+        return cls.from_simple(simple, universe=universe, registry=registry)  # type: ignore
+    except AttributeError as e:
+        raise AttributeError(f"JSON deserialization requires {cls} has a from_simple() class method") from e
```
With a `Protocol` for `from_simple`/`to_simple`, I think we could replace this run-time check with a mypy one.

timj replied:
Sounds like I need to look up protocols. mypy was really annoying in this file.

timj replied:
I've added protocol and removed the ignore. I'm inclined to leave the try/except so you get a more obvious error message. mypy seems happy (mypy didn't use the try/except itself).

TallJimbo reviewed (approved):
Looks good, even though I've left a lot of nitpicky comments - those are mostly musing about how things might or might not work as we'd hope when we start to use them for client/server registry.

Some of those line comments are about whether we should use `registry is not None` in `from_simple` to _assume_ that the incoming form is minimal.  I think that would simplify the code, and it should be just as good at supporting over-the-wire needs, given that client and server should be able to expect when something is using minimal form.  If we do go that route, it may be worth playing with making the minimal versions of things into separate methods, even though that multiplies the number of methods by two, and that's a lot of methods.  But if most classes don't actually have a minimal form, and don't need those methods, maybe that's better?  I'm not sure either way.

Finally, I think you've got the right amount of checking for invalid inputs here, in the sense that they're reasonable things the user might get wrong, but there's no attempt to do really exhaustive schema validation here (which I think is not something we should roll ourselves, if we want it).  I don't actually know enough about developing http interfaces to know if we _should_ want exhaustive schema validation - but if there's a chance that we do, that's another thing we should think about before really settling on this approach.

timj commented:
Regarding schema validation, it seems to me that we'd probably want to look at pydantic for that https://pydantic-docs.helpmanual.io/usage/exporting_models/  -- it has json support. Naively we could make the `to_simple` methods return a `SerializedDatasetRef` (or whatever) that is defined by the pydantic model. The constructor for that would take the json plain-old-data and then we'd call `from_simple` on that `SerializedDatasetRef`. Pydantic would handle all the checks. Of course the other option is to make `DatasetRef` an actual pydantic model but that might be a complicated retrofit.   Supporting `DimensionsRecords` seems a bit harder but maybe we could create a SerializedVisitRecord and SerializedExposureRecord dynamically from the dimensions.yaml.

TallJimbo commented:
I'm mostly reasoning by analogy with Rust's Serde, which sounds pretty similar to pydantic, but if we go in that direction I'd lean towards:
 - `to_pydantic`/`from_pydantic` convert out objects to a corresponding pydantic model instance; may return `self` (and similar for `from`) if the class can easily just be a pydantic model itself;
 - make `to_json`/`from_json` delegate to that;
 - add `to_simple`/`from_simple` that also delegate to that, if it looks like we still have a need for them.

timj committed:
Change DatasetRef simple form to always be dict

Remove int id option because we are sure that in the future
an "origin" will be required.

timj commented on python/lsst/daf/butler/core/datasets/ref.py:
```diff
@@ -152,6 +158,103 @@ def __lt__(self, other: Any) -> bool:
+        # Convert to a dict form
+        as_dict: Dict[str, Any] = {"datasetType": self.datasetType.to_simple(minimal=minimal),
+                                   "dataId": self.dataId.to_simple(),
+                                   "hasParentId": self.hasParentId
```
@TallJimbo I'm fairly certain that `hasParentId` is a left over for when we first changed the handling of components in registry and it was possible to have a component in registry or in datastore depending on how the put had been implemented.

It was added in #269 and I don't see any code that checks the value anywhere.

TallJimbo replied:
I don't recall anything that uses it (I've forgotten the details how we identify component dataset refs), so I'm all for removing if it doesn't break anything.

timj replied:
I removed it and everything is still working...

timj committed:
Remove hasParentId from DatasetRef

It is a setting that is no longer necessary now that registry
no longer deals with components directly and they are an
implementation detail of datastore.

timj committed:
Change region JSON to not match YAML

YAML has the cls field for historically reasons but it's
not actually needed. Simplify by removing it. Must use binhex
string here because JSON does not understand bytes.

timj committed:
Switch DatasetType.from_simple from querying to getting

timj committed:
Use protocols to keep mypy happy with to/from_json methods
